core/src/main/scala/kafka/log/Log.scala

```diff
@@ -417,25 +418,14 @@ class Log(@volatile var dir: File,
    * @return The number of bytes truncated from the segment
    * @throws LogSegmentOffsetOverflowException if the segment contains messages that cause index offset overflow
    */
-  private def recoverSegment(segment: LogSegment, leaderEpochCache: Option[LeaderEpochCache] = None): Int = lock synchronized {
-    val stateManager = new ProducerStateManager(topicPartition, dir, maxProducerIdExpirationMs)
-    stateManager.truncateAndReload(logStartOffset, segment.baseOffset, time.milliseconds)
-    logSegments(stateManager.mapEndOffset, segment.baseOffset).foreach { segment =>
-      val startOffset = math.max(segment.baseOffset, stateManager.mapEndOffset)
-      val fetchDataInfo = segment.read(startOffset, None, Int.MaxValue)
-      if (fetchDataInfo != null)
-        loadProducersFromLog(stateManager, fetchDataInfo.records)
-    }
-    stateManager.updateMapEndOffset(segment.baseOffset)
-
-    // take a snapshot for the first recovered segment to avoid reloading all the segments if we shutdown before we
-    // checkpoint the recovery point
-    stateManager.takeSnapshot()
-    val bytesTruncated = segment.recover(stateManager, leaderEpochCache)
-
+  private def recoverSegment(segment: LogSegment,
+                             leaderEpochCache: Option[LeaderEpochCache] = None): Int = lock synchronized {
+    val producerStateManager = new ProducerStateManager(topicPartition, dir, maxProducerIdExpirationMs)
+    loadProducerState(segment.baseOffset, reloadFromCleanShutdown = false, producerStateManager)
```

Comment:
Hmm.. `loadProducerState` also update the first unstable offset. Since we're trying to keep the recovery logic free of side effects, I'd suggest pulling the call to `updateFirstUnstableOffset()` out. Maybe something like this?

```scala
private def rebuildProducerState(lastOffset: Long,
                                 reloadFromCleanShutdown: Boolean,
                                 producerStateManager: ProducerStateManager) = {
 // load producer state logic except update to first unstable offset
}

private def loadProducerState(lastOffset: Long, 
                              reloadFromCleanShutdown: Boolean) = {
  rebuildProducerState(lastOffset, reloadFromCleanShutdown, producerStateManager)
  updateFirstUnstableOffset()
}
```